src/main/scala/com/fulcrumgenomics/bam/SplitBam.scala

```diff
@@ -55,77 +55,101 @@ object SplitType extends FgBioEnum[SplitType] {
-    val counter  = new SimpleCounter[WriterInfo]()
+    val writerInfoMap: Map[SAMReadGroupRecord, WriterInfo] = {
+      createWriters(header=in.header, splitBy=splitBy).withDefaultValue(unknownBamAndWriter)
+    }
 
     in.foreach { rec =>
-      val info = writers(rec.readGroup)
-      info.writer += rec
-      counter.count(info)
+      val info = writerInfoMap(rec.readGroup)
+      info.count += 1
+      info.writer.write(rec)
       progress.record(rec)
     }
+    progress.logLast()
 
-    writers.values.foreach(_.close())
-    unknownBamAndWriter.close()
     in.safelyClose()
 
-    counter.toSeq sortBy { _._1.name } foreach { case (info, count) =>
-      logger.info(s"Wrote $count records to ${info.bam.getFileName}")
+    writerInfoMap.values.toSeq.sortBy(_.name).distinct.foreach { info =>
+      info.writer.close()
+      logger.info(f"Wrote ${info.count}%,d records to ${info.bam.getFileName}")
     }
 
-    if (counter.countOf(unknownBamAndWriter) == 0) {
+    if (unknownBamAndWriter.count == 0) {
       Files.delete(unknownBamAndWriter.bam)
     }
   }
 
-  /** Stores the path to the BAM and the associated writer */
-  private case class WriterInfo(name: String, bam: PathToBam, writer: SamWriter) extends Closeable{
-    def close(): Unit = writer.close()
+  /** Gets the output path for the writer with a given name. */
+  private[bam] def toOutput(name: String): PathToBam = {
+    val outputDir = output.getParent
+    val prefix    = output.getFileName
+    outputDir.resolve(PathUtil.sanitizeFileName(s"$prefix.$name.bam"))
   }
 
   /** Initializes the writers and returns a map from each read group to a writer. */
   private def createWriters(header: SAMFileHeader, splitBy: SplitType): Map[SAMReadGroupRecord, WriterInfo] = {
-    splitBy match {
-      case SplitType.Library =>
-        header.getReadGroups.toSeq.groupBy { rg => rg.getLibrary }
-          .flatMap { case (library, readGroups) =>
-            val bam = toOutput(library)
-            val writer = SamWriter(bam, header)
-            readGroups.map { rg => rg -> WriterInfo(name=library, bam=bam, writer=writer) }
-          }
-      case SplitType.ReadGroup =>
-        header.getReadGroups.map { rg =>
-          val bam = toOutput(rg.getId)
-          val writer = SamWriter(bam, header)
-          rg -> WriterInfo(name=rg.getId, bam=bam, writer=writer)
-        }.toMap
+    val groups: Map[String, Seq[SAMReadGroupRecord]]  = splitBy match {
+      case SplitType.Library   => header.getReadGroups.toSeq.groupBy { rg => rg.getLibrary }
+      case SplitType.ReadGroup => header.getReadGroups.map(rg => rg.getId -> Seq(rg)).toMap
+    }
+    logger.info(f"Outputting to ${groups.size}%,d files ($splitBy).")
+    groups.flatMap { case (library, readGroups) =>
+      val bam    = toOutput(library)
+      val writer = SamWriter(
+        path   = bam,
+        header = toOutputHeader(header=header, readGroups:_*),
+        async  = !noAsyncWriting
```

Comment:
If `--async-io` is used, then we'll have a thread per `SamWriter`, which for 10K read groups is too many for some systems.  Even when we up the thread-per-process limit, each thread will buffer, and so we use a too much memory.   It runs fast enough if we just disable async io for writing.